### PR TITLE
Correlation metrics

### PR DESCRIPTION
**Pull request by tildechris**
**Prerequisites**:

- [x] Unless it is a hotfix it should be merged against the `dev` branch
- [x] Branch is up-to-date with the branch to be merged with, i.e. `dev`
- [x] Build is successful
- [x] Code is cleaned up and formatted

### Summary

Add a utility method that generates overall r^2, and then over a +/- 28 day period max r^2, lag at max and a series such that the r^2 vs. lag can be plotted.

### CODE REVIEW

**Comment by tildechris** [comment]
The regression library does add heft to the bundled js.  This logic should be moved to the server when possible.

**Comment by sgratzl** [comment]
> The regression library does add heft to the bundled js. This logic should be moved to the server when possible.

do you mean bundle size wise? as far as I can see it is actually a quite small library

File: package.json [hunk 1]
```diff
@@ -100,7 +100,7 @@
     "mapbox-gl": "^1.13.0",
     "marked": "^2.0.1",
     "mini-css-extract-plugin": "^1.3.9",
-    "node-fetch": "^2.6.1",
```
Comment (sgratzl) [inline package.json:103]: why did you remove this dependency?

File: src/data/testData.js [hunk 2]
```diff
@@ -0,0 +1,3151 @@
+export let deathsTestData = [
```
Comment (sgratzl) [inline src/data/testData.js:1]: what about moving this test data to some directory like `data/__tests__/testData.js` to make it clear it is for tests only?

File: src/data/utils.js [hunk 3]
```diff
@@ -38,3 +39,83 @@ export function parseAPITime(v) {
...
+/**
+ * Generates R^2 metrics for lags between -28 and 28 days.
+ *
+ * For lags between 0 and 28 lag b backwards with respect to a.  For -28 to -1 lag a with
+ * respect to b.
+ *
+ * For each lag, the input is a window of length(a)-28, such that the number of values at
+ * each lag is the same.
+ *
+ * @param {Number[]} a
```
Comment (sgratzl) [inline src/data/utils.js:60]: prefer `number[]` instead of `Number[]`

File: src/data/utils.js [hunk 4]
```diff
@@ -38,3 +39,83 @@ export function parseAPITime(v) {
...
+ * @param {Number[]} b
```
Comment (sgratzl) [inline src/data/utils.js:61]: prefer also adding a `@returns {type}` hint

File: src/data/utils.js [hunk 5]
```diff
@@ -38,3 +39,83 @@ export function parseAPITime(v) {
...
+ */
+function generateLags(a, b) {
+  let lag = 28;
```
Comment (sgratzl) [inline src/data/utils.js:64]: prefer `const` instead of `let` when possible

File: src/data/utils.js [hunk 6]
```diff
@@ -38,3 +39,83 @@ export function parseAPITime(v) {
...
+    return [x, b[i]];
+  });
+}
+
+/**
+ * Compute 28-day correlation metrics for a response variable given an explanatory variable.
+ *
+ * @param {import('./fetchData').EpiDataRow[]} response
+ * @param {import('./fetchData').EpiDataRow[]} explanatory
+ * @param {CorrelationMetric} metric
```
Comment (sgratzl) [inline src/data/utils.js:99]: isn't that `@returns`?

File: src/data/utils.js [hunk 7]
```diff
@@ -38,3 +39,83 @@ export function parseAPITime(v) {
...
+ */
+export function generateCorrelationMetrics(response, explanatory) {
+  let response_values = response.map((row) => {
```
Comment (sgratzl) [inline src/data/utils.js:102]: prefer inline arrow function such as `(row) => row.value`, easier to read for such simple functions

File: src/data/utils.js [hunk 8]
```diff
@@ -38,3 +39,83 @@ export function parseAPITime(v) {
...
+  });
+  let explanatory_values = explanatory.map((row) => {
+    return row.value;
+  });
+
+  let model = linear(zip(response_values, explanatory_values));
+  let lags = generateLags(response_values, explanatory_values);
+  let max = lags.reduce((acc, i) => {
+    return i.r2 > acc.r2 ? i : acc;
+  });
```
Comment (sgratzl) [inline src/data/utils.js:113]: missing reduce initial argument?

**Review by sgratzl — COMMENTED** [review]
in general I prefer using `const` over `let` when possible. This gives also the compiler more hints about how to optimize the code

File: package.json (hunk 1 above)
Comment (tildechris) [reply package.json:103]: I was using it initially in the unit test to use the API, then reverted my changes when I extracted test data.  I didn't notice that it was used elsewhere.  Will revert.

File: src/data/testData.js (hunk 2 above)
Comment (tildechris) [reply src/data/testData.js:1]: Good suggestion.  I didn't see an existing spot.

File: src/data/utils.spec.js [hunk 9]
```diff
@@ -22,3 +23,18 @@ describe('parseAPITime', () => {
...
+
+describe('correlationMetrics', () => {
+  test('Deaths vs. CLI correlations should equal what was published in the blog post', async () => {
+    let expected_metrics = {
+      r2At0: 0.43,
+      lagAtMaxR2: 20,
+      r2AtMaxR2: 0.81,
+    };
+    let actual_metrics = generateCorrelationMetrics(deathsTestData, cliTestData);
+    expect(actual_metrics.r2At0).toEqual(expected_metrics.r2At0);
```
Comment (sgratzl) [inline src/data/utils.spec.js:35]: an alternative method to defining the values directly is to use snapshot testing: https://jestjs.io/docs/snapshot-testing

File: src/data/utils.js [hunk 10]
```diff
@@ -38,3 +39,87 @@ export function parseAPITime(v) {
...
+    lags.push({ lag: -1 * i, r2: model.r2 });
+  }
+  return lags;
+}
+
+/**
+ * Do a pair-wise combination of the elements from a and b arrays.
+ * @param {number[]} a
+ * @param {number[]} b
+ * @returns {number[]}
```
Comment (sgratzl) [inline src/data/utils.js:94]: isn't it `[number, number][]` a tuple of two numbers?

**Comment by tildechris** [comment]
> > The regression library does add heft to the bundled js. This logic should be moved to the server when possible.
> 
> do you mean bundle size wise? as far as I can see it is actually a quite small library

I thought I saw a number of transitive dependencies, but you're right it's pretty lightweight.  Still would make sense to do server side, however.

File: src/data/utils.spec.js (hunk 9 above)
Comment (tildechris) [reply src/data/utils.spec.js:35]: That's a pretty neat feature.  Here's however, I wanted to be a little looser about the assertion, specifically in lags array that is returned.  Currently it's not sorted with respect to lag number, so if the ordering changes in the future, I didn't want to break the test.

File: src/data/utils.js (hunk 8 above)
Comment (tildechris) [reply src/data/utils.js:117]: [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce) says that "If no initialValue is supplied, the first element in the array will be used as the initial accumulator value and skipped as currentValue".

**Review by sgratzl — APPROVED** [review]
👍
